src/peft/tuners/delora/layer.py

```diff
@@ -0,0 +1,630 @@
+            return
+
+        for active_adapter in adapter_names:
+            if active_adapter in self.delora_A.keys():
+                base_layer = self.get_base_layer()
+                delta_weight = (
+                    self.get_delta_weight(active_adapter)
+                    .detach()
+                    .to(dtype=base_layer.weight.dtype, device=base_layer.weight.device)
+                )
+                with torch.no_grad():
+                    if safe_merge:
+                        orig_weights = base_layer.weight.data.clone()
+                        orig_weights = orig_weights + delta_weight
+
+                        if not torch.isfinite(orig_weights).all():
+                            raise ValueError(
+                                f"NaNs detected in merged weights for adapter {active_adapter}; aborting merge"
+                            )
+                        
+                        base_layer.weight.data = orig_weights
+                    else:
+                        base_layer.weight.data.add_(delta_weight)
+
+                self.merged_adapters.append(active_adapter)
+
+    def unmerge(self) -> None:
+        """Unmerge all merged adapter layers from the base weights."""
+        if not self.merged:
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+        while len(self.merged_adapters) > 0:
+            active_adapter = self.merged_adapters.pop()
+            if active_adapter in self.delora_A.keys():
+                with torch.no_grad():
+                    delta_weight = self.get_delta_weight(active_adapter)
+                    self.get_base_layer().weight.data.sub_(delta_weight)
+        # If we had subtracted initial delta(s) at setup, restore them now to get back the original base weights
+        # This should not be done for a model that has just been loaded from a checkpoint, as the state is already correct.
+        if self.use_residual_init and len(self._initials_applied) > 0:
+            w = self.get_base_layer().weight
+            for adapter in list(self._initials_applied):
+                if (
+                    (adapter in self.delora_initial_A)
+                    and (adapter in self.delora_initial_B)
+                    and (adapter in self.delora_initial_w_norm)
+                ):
+                    A0 = self.delora_initial_A[adapter].to(dtype=w.dtype, device=w.device)
+                    B0 = self.delora_initial_B[adapter].to(dtype=w.dtype, device=w.device)
+                    alpha0 = self.delora_initial_alpha[adapter].to(dtype=w.dtype, device=w.device).squeeze(0)
+                    wnorm0 = self.delora_initial_w_norm[adapter].to(dtype=w.dtype, device=w.device)
+                    delta0 = self._compute_delta_from_params(A0, B0, alpha0, self.r[adapter], wnorm0)
+                    self.get_base_layer().weight.data = self.get_base_layer().weight.data + delta0
+                self._initials_applied.discard(adapter)
```

Comment:
Since we checked earlier that `adapter` is in `_initials_applied`, there is no need to use `discard`, you can use `remove`.